documentation/add/04_mpi_framework.md

```diff
@@ -109,185 +146,232 @@ The main methods and properties (M/P) of **JobExecutor** involved in a job contr
+
+Independent workers may need to report to the other workers some special conditions occurring during their execution. 
+Such conditions currently result in a failure (an interrupt is raised at code execution). 
+We use Persistent messages aka Interrupt messages to distribute information about such states of the system as parallel interrupts, as MPI interrupts used for these purposes in standard MPI frameworks are not available in all parallel frameworks. 
+Such a message, received by a worker from other worker, block all further communication with this worker. 
+All requests for information from this worker result in the persistent message (*failure*). 
+This state holds until special command (*clear_all* above) is executed by the receiving worker.
 
 #### Messages types.
 
-Different purposes can be best served using different types of messages. There are 3 purposes, differentiated in Horace.
+Different operations can be handled using different types of messages. There are 3 main types, differentiated in Horace.
 
-1. logging, diagnostics and informing user about progress of a job.
+1. Logging, diagnostics and informing user about progress of a job.
 2. Interprocess data exchange, necessary for performing a particular user job and
-3. The information about error or failure.
+3. Information about error or failure.
 
 
-Correspondent types of messages created to serve these purposes:
+Corresponding types of messages created to serve these purposes:
 
-1. Non-blocking (transient)  messages. Informing user about a job progress is important task, but user is interested only in the final value. Different independent task should not be synchronous with regard to log messages as only some average progress is required. Non-blocking messages are best suited for such kind of task. If more then one non-blocking message is present in the MPI messages queue, the client receives only the last message.
-2. Blocking messages are necessary to transfer information between independent workers when the result of calculations on a single node depends on the results on other nodes.
-3. Persistent messages necessary to indicate some critical conditions or error states, as, for example, parallel interrupts available in MPI are not available in file-based messages system (*ClusterHerbert*) and not accessible for Matlab MPI (*ClusterParpool*). The persistent messages used to carry out such information. 
+1. Non-blocking (transient)  messages. Informing user about a job progress is an important task, but often the user is interested only in the final value.
+   Different independent tasks should not be synchronous with regard to log messages as only some average progress is required. 
+   Non-blocking messages are best suited for such kind of task. 
+   If more then one non-blocking message is present in the MPI messages queue, the client receives only the last message.
```

Comment:
Does this worry us? Can we improve this?

Reply:
this is intentional and took a lot of efforts to implement (in some cases)

Reply:
Why is it intentional? What is the advantage?

Reply:
if the messages are synchronous, one need to synchronize on them. This synchronization is artificial, as independent processes do not need to know each others progress and synchronization takes time. The purpose of asynchronous messages is mainly logging and in logging you are interested in the last state of an interdependent worker only.

Reply:
If we are trying to trace the execution of a process logging all messages would be useful: the final state will not necessarily contain the information about what triggered a code path that led to failure. 
These are async in as much as we don't need to block the main process execution, but they are still important.
The messages (can) carry information about which process they are from and a timestamp so it doesn't matter if they are interlaced, or even out-of-order in the log files.